internal/app/init/shutdown.go

```diff
@@ -0,0 +1,65 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package main
+
+import (
+	"os"
+
+	"github.com/mdlayher/genetlink"
+	"github.com/mdlayher/netlink"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// See https://github.com/torvalds/linux/blob/master/drivers/acpi/event.c
+	acpiGenlFamilyName     = "acpi_event"
+	acpiGenlMcastGroupName = "acpi_mc_group"
+)
+
+func listenForPowerButton() (err error) {
+	// Get the acpi_event family.
+
+	genconn, err := genetlink.Dial(nil)
+	if err != nil {
+		return err
+	}
+	// nolint: errcheck
+	defer genconn.Close()
+	var f genetlink.Family
+	if f, err = genconn.GetFamily(acpiGenlFamilyName); os.IsNotExist(err) {
+		return errors.Wrap(err, acpiGenlFamilyName+" not available")
+	}
+
+	// Listen for ACPI event.
+
+	conn, err := netlink.Dial(unix.NETLINK_GENERIC, nil)
+	if err != nil {
+		return err
+	}
+	// nolint: errcheck
+	defer conn.Close()
+	var id uint32
+	for _, group := range f.Groups {
+		if group.Name == acpiGenlMcastGroupName {
+			id = group.ID
+		}
+	}
+	if err = conn.JoinGroup(id); err != nil {
+		return err
+	}
+	msgs, err := conn.Receive()
+	if err != nil {
+		return err
+	}
+	if len(msgs) > 0 {
+		// TODO(andrewrynhard): Stop all running containerd tasks.
```

Comment:
Is this mostly a matter of instantiating a new osd client and triggering the necessary shutdown work there?

Reply:
I'm imagining that we would eventually use the `osd` API to shutdown a node (See #376). I think I'd like to implement the shutdown API soon. So yes to your answer but the blocker is that API doesn't exist yet.